cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -0,0 +1,134 @@
+        String gitFileHash = repositoryManager.getFileHash(relativeFilePath, pushedCommit);
+        return new PushAppResponse(appToPush.getName(), appToPush.getAppVersion(), gitFileHash);
+      } catch (Exception e) {
+        throw new PushFailureException("Failed to fetch push details", e);
+      }
+    }
+  }
+
+  private String writeAppDetailAndPush(RepositoryManager repositoryManager,
+                                       ApplicationDetail appToPush,
+                                       CommitMeta commitDetails)
+    throws PushFailureException, NoChangesToPushException {
+    Path appWrittenPath;
+    try {
+      String appName = appToPush.getName();
+      // Symlinks will be followed but if the path points to an outside directory it will fail
+      appWrittenPath = getAndValidateAppConfigRelativePath(repositoryManager, appName);
+      // Opens the file for writing, creating the file if it doesn't exist,
+      // or truncating an existing regular-file to a size of 0
+      Path filePathToWrite = repositoryManager.getRepositoryRoot().resolve(appWrittenPath);
+      try (FileWriter writer = new FileWriter(filePathToWrite.toString())) {
+        GSON.toJson(appToPush, writer);
+      }
+
+      LOG.debug("Written application configs for {} in file {}", appName, appWrittenPath);
+    } catch (Exception e) {
+      throw new PushFailureException("Failed to write app configs", e);
+    }
+
+    try {
+      String pushedCommit = repositoryManager.commitAndPush(commitDetails, Collections.singletonList(appWrittenPath));
+      LOG.info("Pushed application configs for : {}", appToPush);
+      return pushedCommit;
+    } catch (UnexpectedRepositoryChangesException | GitAPIException e) {
+      throw new PushFailureException("Failed to push to git", e);
+    }
+  }
+
+  /**
+   * Generates the config file relative path in git root directory based on application name.
+   * It also validates if the resolved path ( considering symlinks) exists outside git repository
+   * or not.
+   *
+   * @param repositoryManager the RepositoryManager
+   * @param applicationName the name of this application
+   * @return A valid application config file path
+   */
+  private Path getAndValidateAppConfigRelativePath(RepositoryManager repositoryManager,
+                                                   String applicationName)
+    throws InvalidPathInSourceControl, IOException {
+    String fileName = String.format("%s.json", applicationName);
+    Path filePath = repositoryManager.getBasePath().resolve(fileName);
+    if (Files.exists(filePath)) {
+      Path resolvedPath = filePath.toRealPath();
+      Path resolvedRepoRootPath = repositoryManager.getRepositoryRoot().toRealPath();
+      if (!resolvedPath.startsWith(resolvedRepoRootPath)) {
+        throw new InvalidPathInSourceControl(String.format("Config file %s pointing to invalid path %s", filePath,
```

Comment:
from the method name I thought this would be checking if the path is a directory and not something about following symlinks.

I'm not even sure we want to do this symlink resolution, it's a weird scenario. For example, what if you have multiple different pipelines that are symlinked to the same file? I think it's better just to fail if we encounter one.

Reply:
Also should check if the path is a directory, then we can't write to it.

Reply:
I feel this check is necessary because some systems like macOS does symlinks already, it symlinks `/tmp` to `/private/tmp`: https://apple.stackexchange.com/questions/1043/why-is-tmp-a-symlink-to-private-tmp

We need to see if the path is actually pointed to the repo directory

Reply:
Added the directory check